File: SmartHomeTEC_API/Models/Device.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace SmartHomeTEC_API.Models
{
    [Table("Device")] // Especifica el nombre exacto de la tabla en la base de datos
    public class Device
    {
        [Key]
        [Required]
        public required string SerialNumber { get; set; }

        [Required]
        [Range(0, double.MaxValue, ErrorMessage = "El precio debe ser un número no negativo.")]
        public required decimal Price { get; set; }

        [Required]
        [JsonConverter(typeof(JsonStringEnumConverter))]
        public required DeviceState State { get; set; }

        [Required]
        public required string Brand { get; set; }

        public int? AmountAvailable { get; set; } // Campo opcional

        [Range(0, double.MaxValue, ErrorMessage = "El consumo eléctrico debe ser un valor positivo.")]
        public double? ElectricalConsumption { get; set; }

        [Required]
        public required string Name { get; set; }

        [Required]
        public required string Description { get; set; }

        // Foreign Key a DeviceType
        [Required]
        [ForeignKey("DeviceType")]
        public required string DeviceTypeName { get; set; }

        [JsonIgnore]
        public DeviceType? DeviceType { get; set; } // Navegación

        // Foreign Key a Distributor
        [ForeignKey("Distributor")]
        public string? LegalNum { get; set; }

        [JsonIgnore]
        public Distributor? Distributor { get; set; } // Navegación

        // Propiedad de Navegación hacia Order
        [JsonIgnore]
        public Order? Order { get; set; }

    }

    public enum DeviceState
    {
        AdminRegistered,
        StoreAvailable,
        Local
    }
}

File: smart_home_tec_app/lib/JSONmodels/device.dart
import 'dart:convert';

Device deviceFromJson(String str) => Device.fromJson(json.decode(str));

String deviceToJson(Device data) => json.encode(data.toJson());

class Device {
    final int serialNumber;
    final int price;
    final String state;
    final String brand;
    final int amountAvailable;
    final String electricalConsumption;
    final String name;
    final String description;
    final String deviceTypeName;
    final int legalNum;

    Device({
        required this.serialNumber,
        required this.price,
        required this.state,
        required this.brand,
        required this.amountAvailable,
        required this.electricalConsumption,
        required this.name,
        required this.description,
        required this.deviceTypeName,
        required this.legalNum,
    });

    factory Device.fromJson(Map<String, dynamic> json) => Device(
        serialNumber: json["serialNumber"],
        price: json["price"],
        state: json["state"],
        brand: json["brand"],
        amountAvailable: json["amountAvailable"],
        electricalConsumption: json["electricalConsumption"],
        name: json["name"],
        description: json["description"],
        deviceTypeName: json["deviceTypeName"],
        legalNum: json["legalNum"],
    );

    Map<String, dynamic> toJson() => {
        "serialNumber": serialNumber,
        "price": price,
        "state": state,
        "brand": brand,
        "amountAvailable": amountAvailable,
        "electricalConsumption": electricalConsumption,
        "name": name,
        "description": description,
        "deviceTypeName": deviceTypeName,
        "legalNum": legalNum,
    };
}
